File: src/main/java/net/andrecarbajal/naviMusic/audio/AudioResultHandler.java
package net.andrecarbajal.naviMusic.audio;

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import dev.lavalink.youtube.YoutubeAudioSourceManager;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import net.andrecarbajal.naviMusic.dto.VideoInfo;
import net.andrecarbajal.naviMusic.dto.response.Response;
import net.andrecarbajal.naviMusic.dto.response.RichResponse;
import net.andrecarbajal.naviMusic.util.URLUtils;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;


@Slf4j
public class AudioResultHandler implements AudioLoadResultHandler {
    public AudioResultHandler(MusicService musicService, Guild guild, Member member) {
        this.musicService = musicService;
        this.guild = guild;
        this.member = member;
    }

    @Getter
    private Response response;

    private final MusicService musicService;
    private final Guild guild;
    private final Member member;

    @Override
    public void trackLoaded(AudioTrack track) {
        String uri = track.getInfo().uri;

        RichResponse r = new RichResponse();

        r.setColor(new Color(0, 51, 102));
        r.setTitle("Song added to queue");
        r.setText(String.format("[%s](%s) de `%s`", track.getInfo().title.trim(), uri, track.getInfo().author));

        List<MessageEmbed.Field> fields = new ArrayList<>();
        fields.add(new MessageEmbed.Field("Duration", new VideoInfo(track.getInfo()).durationToReadable(), true));

        int size = musicService.getGuildMusicManager(guild).getScheduler().getQueueSize()+1;
        fields.add(new MessageEmbed.Field("In queue", String.format(size==1?"%d song":"%d songs", size), true));

        r.setFields(fields);

        r.setFooter(new RichResponse.Footer(String.format("Added by %s", member.getEffectiveName()), member.getEffectiveAvatarUrl()));

        if (track.getSourceManager() instanceof YoutubeAudioSourceManager) {
            URLUtils.getURLParam(uri, "v").ifPresent(s -> r.setThumbnail(String.format("https://img.youtube.com/vi/%s/maxresdefault.jpg", s)));
        }

        response = r;
        musicService.play(guild, musicService.getGuildMusicManager(guild), track, member);
    }

    @Override
    public void playlistLoaded(AudioPlaylist playlist) {
        AudioTrack firstTrack = playlist.getSelectedTrack();

        if (firstTrack == null) {
            firstTrack = playlist.getTracks().getFirst();
        }

        if (playlist.isSearchResult()) {
            trackLoaded(firstTrack);
            return;
        }

        RichResponse r = new RichResponse();

        r.setColor(new Color(0, 51, 102));
        r.setTitle("Playlist added to queue");

        r.setText(String.format("%s", playlist.getName()));

        int playlistSize = playlist.getTracks().size();
        List<MessageEmbed.Field> fields = new ArrayList<>();
        fields.add(new MessageEmbed.Field("Songs added", String.valueOf(playlistSize), true));

        int size = musicService.getGuildMusicManager(guild).getScheduler().getQueueSize()+playlistSize;
        fields.add(new MessageEmbed.Field("In queue", String.format(size==1?"%d song":"%d songs", size), true));

        r.setFields(fields);

        r.setFooter(new RichResponse.Footer(String.format("Added by %s", member.getEffectiveName()), member.getEffectiveAvatarUrl()));

        if (firstTrack.getSourceManager() instanceof YoutubeAudioSourceManager) {
            URLUtils.getURLParam(firstTrack.getInfo().uri, "v").ifPresent(s -> r.setThumbnail(String.format("https://img.youtube.com/vi/%s/maxresdefault.jpg", s)));
        }

        response = r;
        musicService.playPlaylist(guild, musicService.getGuildMusicManager(guild), playlist, member);
    }

    @Override
    public void noMatches() {
        RichResponse r = new RichResponse();

        r.setType(Response.Type.USER_ERROR);
        r.setText("Nothing found");
        response=r;
    }

    @Override
    public void loadFailed(FriendlyException exception) {
        log.error("Error loading track", exception);

        RichResponse r = new RichResponse();
        r.setType(Response.Type.ERROR);
        r.setText("Internal error");
        response=r;
    }
}

File: src/main/kotlin/net/andrecarbajal/naviMusic/audio/resultHandler/AudioResultHandler.kt
package net.andrecarbajal.naviMusic.audio.resultHandler

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import dev.lavalink.youtube.YoutubeAudioSourceManager
import net.andrecarbajal.naviMusic.audio.MusicService
import net.andrecarbajal.naviMusic.dto.VideoInfo
import net.andrecarbajal.naviMusic.dto.response.Response
import net.andrecarbajal.naviMusic.dto.response.RichResponse
import net.andrecarbajal.naviMusic.util.URLUtils
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import org.slf4j.LoggerFactory

class AudioResultHandler(
    private val musicService: MusicService,
    private val guild: Guild,
    private val member: Member,
    private val event: SlashCommandInteractionEvent? = null
) : AudioLoadResultHandler {

    private val log = LoggerFactory.getLogger(AudioResultHandler::class.java)
    var response: Response? = null
        private set

    override fun trackLoaded(track: AudioTrack) {
        val trackInfo = track.info
        val size = musicService.getGuildMusicManager(guild).scheduler.getQueueSize() + 1

        val richResponse = RichResponse(
            title = "Song added to queue",
            text = "[${trackInfo.title.trim()}](${trackInfo.uri}) de `${trackInfo.author}`",
            fields = listOf(
                MessageEmbed.Field("Duration", VideoInfo(trackInfo).durationToReadable(), true),
                MessageEmbed.Field("In queue", if (size == 1) "1 song" else "$size songs", true)
            ),
            footer = RichResponse.Footer(
                text = "Added by ${member.effectiveName}", imageUrl = member.effectiveAvatarUrl
            )
        )

        if (track.sourceManager is YoutubeAudioSourceManager) {
            URLUtils.getURLParam(trackInfo.uri, "v").ifPresent { s ->
                richResponse.thumbnail = "https://img.youtube.com/vi/$s/maxresdefault.jpg"
            }
        }

        response = richResponse
        event?.let { richResponse.editReply(it) }
        musicService.play(musicService.getGuildMusicManager(guild), track, member)
    }

    override fun playlistLoaded(playlist: AudioPlaylist) {
        val firstTrack = playlist.selectedTrack ?: playlist.tracks.firstOrNull() ?: return

        if (playlist.isSearchResult) {
            trackLoaded(firstTrack)
            return
        }

        val playlistSize = playlist.tracks.size
        val size = musicService.getGuildMusicManager(guild).scheduler.getQueueSize() + playlistSize

        val richResponse = RichResponse(
            title = "Playlist added to queue", text = playlist.name, fields = listOf(
                MessageEmbed.Field("Songs added", playlistSize.toString(), true),
                MessageEmbed.Field("In queue", if (size == 1) "1 song" else "$size songs", true)
            ), footer = RichResponse.Footer(
                text = "Added by ${member.effectiveName}", imageUrl = member.effectiveAvatarUrl
            )
        )

        if (firstTrack.sourceManager is YoutubeAudioSourceManager) {
            URLUtils.getURLParam(firstTrack.info.uri, "v").ifPresent { s ->
                richResponse.thumbnail = "https://img.youtube.com/vi/$s/maxresdefault.jpg"
            }
        }

        response = richResponse
        event?.let { richResponse.editReply(it) }
        musicService.playPlaylist(musicService.getGuildMusicManager(guild), playlist, member)
    }

    override fun noMatches() {
        val richResponse = RichResponse(
            type = Response.Type.USER_ERROR, text = "Nothing found"
        )
        response = richResponse
        event?.let { richResponse.editReply(it) }
    }

    override fun loadFailed(exception: FriendlyException) {
        log.error("Error loading track", exception)
        val richResponse = RichResponse(
            type = Response.Type.ERROR, text = "Internal error"
        )
        response = richResponse
        event?.let { richResponse.editReply(it) }
    }
}
